contracts/pair_anchor/src/contract.rs

```diff
@@ -0,0 +1,839 @@
+///
+/// * **to** is an object of type [`Option<Addr>`]. Sets the recipient of the swap operation.
+/// NOTE - the address that wants to swap should approve the pair contract to pull the offer token.
+#[allow(clippy::too_many_arguments)]
+pub fn swap(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    sender: Addr,
+    offer_asset: Asset,
+    belief_price: Option<Decimal>,
+    max_spread: Option<Decimal>,
+    to: Option<Addr>,
+) -> Result<Response, ContractError> {
+    offer_asset.assert_sent_native_token_balance(&info)?;
+
+    let config: Config = CONFIG.load(deps.storage)?;
+
+    // If the asset balance is already increased, we should subtract the user deposit from the pool amount
+    let pools: Vec<Asset> = config
+        .pair_info
+        .query_pools(&deps.querier, env.clone().contract.address)?
+        .to_vec();
+
+    let offer_pool: Asset;
+    let ask_pool: Asset;
+
+    if offer_asset.info.equal(&pools[0].info) {
+        offer_pool = pools[0].clone();
+        ask_pool = pools[1].clone();
+    } else if offer_asset.info.equal(&pools[1].info) {
+        offer_pool = pools[1].clone();
+        ask_pool = pools[0].clone();
+    } else {
+        return Err(ContractError::AssetMismatch {});
+    }
+
+    if ask_pool.amount > Uint128::from(0u128) {
+        // send to community? Otherwise it would be sent to the swapper
+    }
+
+    let mut messages: Vec<CosmosMsg> = vec![];
+
+    match offer_asset.info {
+        AssetInfo::Token { contract_addr, .. } => {
+            messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: contract_addr.to_string(),
+                funds: vec![],
+                msg: to_binary(&Cw20ExecuteMsg::Send {
+                    contract: config.anchor_market_addr.to_string(),
+                    amount: offer_asset.amount,
+                    msg: to_binary(&AnchorExecuteMsg::RedeemStable {})?,
+                })?,
+            }))
+        }
+
+        AssetInfo::NativeToken { denom, .. } => {
+            let amount = offer_asset.amount;
+            let asset = Asset {
+                info: AssetInfo::NativeToken {
+                    denom: denom.clone(),
+                },
+                amount,
+            };
+            let amount = amount.checked_sub(asset.compute_tax(&deps.querier)?)?;
+
+            messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: config.anchor_market_addr.to_string(),
+                funds: vec![Coin { denom, amount }],
+                msg: to_binary(&AnchorExecuteMsg::DepositStable {})?,
+            }));
+        }
+    }
+
+    let receiver = to.unwrap_or_else(|| sender.clone());
+
+    messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: env.contract.address.to_string(),
+        funds: vec![],
+        msg: to_binary(&ExecuteMsg::AssertAndSend {
+            offer_asset: Asset {
```

Comment:
done

Reply:
If I change it I get this: use of partially moved value: `offer_asset`
partial move occurs because value has type `std::string::String`, which does not implement the `Copy` traitrustc[E0382](https://doc.rust-lang.org/error-index.html#E0382)

Reply:
```offer_asset: offer_asset.clone()```

Reply:
I get that error when Im using the clone()